Metal Decoder Inference, K-Quants added, Sync'd with latest GGML

RRUK01 opened the pull request:
Hey @ggerganov, excellent project! I've been tinkering with it for the past couple of months and wanted to contribute something back.

It's nothing major but I think I've got the k_quants working. I've followed the llama implementation in terms of keeping it behind a flag (WHISPER_K_QUANTS) which is on by default. 

I've tested it for Q2_K - Q6_K for tiny -> large and it's working across all combinations except for the tiny models where it
the model fails to evaluate.

I'm new to contributing to open source (this is my first ever pull request) and C++ so if I've made any mistakes or you have any suggestions just let me know!

New model sizes:

| model  |  k_quant | size (mb) |
|--------|----------|-----|
| base | Q2_K |  29.9|
| base | Q3_K | 37.1 |
| base | Q4_K | 46.5 |
| base | Q5_K | 55.3 |
| base | Q6_K | 64.7 |
| small | Q2_K | 89.7 |
| small | Q3_K | 113.8 |
| small | Q4_K | 145.5 |
| small | Q5_K | 175.2 |
| small | Q6_K | 206.8 |
| medium | Q2_K | 266.9 |
| medium | Q3_K | 343.9 |
| medium | Q4_K | 444.5 |
| medium | Q5_K | 539.2 |
| medium | Q6_K | 639.9 |
| large | Q2_K | 529.3 |
| large | Q3_K | 685.1 |
| large | Q4_K | 888.9 |
| large | Q5_K | 1080.8 |
| large | Q6_K | 1284.6 |

bobqianic commented:
@RRUK01 Hi, could you please check if pull request #1148 resolves the issue with evaluating the tiny models? I suspect that the issue may be due to an error in the generation of the log-mel spectrogram.

RRUK01 commented:
> @RRUK01 Hi, could you please check if pull request #1148 resolves the issue with evaluating the tiny models? I suspect that the issue may be due to an error in the generation of the log-mel spectrogram.

Hey, I realised in the end that the tiny model doesn't work as it has a n_audio_state = 384 and k_quants at the time required a layer to be a multiple of 256 (base->large all are) as this is the 'block size'. There has been an update since and k_quants can now be done with a block size of 64 so it should be quite easy to update this pull request if anyone is requiring a k_quant'd tiny model. 

Also nice work on the mel_spec, you've done some really good work there!

RRUK01 commented:
**Update**

Added a few things to this pull request:
- Decoder inference on Metal  #1047
- Sync'd with latest GGML updates

I had high hopes for the Metal inference but unfortunately it's fallen a bit flat, it might be that my implementation isn't optimal or there's an error in my code but on testing Metal inference runs slower than CPU (see examples below). I'm wondering it it's because the matrix's involved are relatively small compared with Llama and as such the overhead from using the GPU isn't offset by it's speed, especially compared with the speed up we're already getting from the Accelerate framework. The gap does appear to close on large models but not enough for it to overtake the CPU. 

In implementing this I also sync'd up Whisper's GGML with Llama's. Non-metal inference runs as normal. 

For anyone wanting to test/use this just be aware that to run on the GPU I had to make a small model change and convert the decoder.positional_embedding layer (dpe) to FP16 as GGML Metal doesn't support FP32 currently. I've made a change to the model loading code so that when WHISPER_USE_METAL=1 it expects dpe to be FP16 and when WHISPER_USE_METAL=0 it expects dpe to be FP32. I've also updated the quantization code so that when you quantize a model it will convert dpe to FP16. Be aware thought that because of these changes models quantized from this pull request won't run when WHISPER_USE_METAL=0 unless you change the non-metal dpe to expect FP16. 

The below is from running this on a M2 Macbook Pro 16GB

| Model | Decode time per run in ms (CPU) | Decode time per run in ms (Metal) |
|----|--|------|
| Small (Q5_K) | 4.24 | 7.00 |
| Medium (Q5_K) | 11.47 | 14.43  |
| Large (Q5_K) | 25.80 | 28.37 |

See below for details of some test runs. One other thing to note is that on the smaller models (<= small) the output is slightly difference, the GPU output seems to really like exclamation points for some reason. 

<details><summary>Small Q5_K CPU</summary>

```
whisper_init_from_file_no_state: loading model from 'models/ggml-small_q5_k.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 768
whisper_model_load: n_audio_head  = 12
whisper_model_load: n_audio_layer = 12
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 768
whisper_model_load: n_text_head   = 12
whisper_model_load: n_text_layer  = 12
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 13
whisper_model_load: qntvr         = 2
whisper_model_load: type          = 3
whisper_model_load: mem required  =  453.00 MB (+   16.00 MB per decoder)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: model ctx     =  166.10 MB
whisper_model_load: model size    =  165.87 MB
whisper_init_state: kv self size  =   15.75 MB
whisper_init_state: kv cross size =   52.73 MB

system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | COREML = 0 | OPENVINO = 0 | METAL = 0 | K_QUANTS = 1 | 

main: processing './samples/jfk.wav' (176000 samples, 11.0 sec), 4 threads, 1 processors, lang = en, task = transcribe, timestamps = 1 ...


[00:00:00.000 --> 00:00:11.000]   And so, my fellow Americans, ask not what your country can do for you, ask what you can do for your country.


whisper_print_timings:     load time =    97.75 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    33.77 ms
whisper_print_timings:   sample time =    10.94 ms /    28 runs (    0.39 ms per run)
whisper_print_timings:   encode time =   986.92 ms /     1 runs (  986.92 ms per run)
whisper_print_timings:   decode time =   118.77 ms /    28 runs (    4.24 ms per run)
whisper_print_timings:    total time =  1258.26 ms

``` 
</details>

<details><summary>Small Q5_K GPU</summary>

```
whisper_init_from_file_no_state: loading model from 'models/ggml-small_q5_k.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 768
whisper_model_load: n_audio_head  = 12
whisper_model_load: n_audio_layer = 12
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 768
whisper_model_load: n_text_head   = 12
whisper_model_load: n_text_layer  = 12
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 13
whisper_model_load: qntvr         = 2
whisper_model_load: type          = 3
whisper_model_load: mem required  =  453.00 MB (+   16.00 MB per decoder)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: model ctx     =  166.10 MB
whisper_model_load: model size    =  165.87 MB
whisper_init_state: kv self size  =   15.75 MB
whisper_init_state: kv cross size =   52.73 MB
ggml_metal_init: allocating
ggml_metal_init: using MPS
ggml_metal_init: loaded kernel_add                            0x13881aa60
ggml_metal_init: loaded kernel_add_row                        0x13881cc20
ggml_metal_init: loaded kernel_mul                            0x13881c080
ggml_metal_init: loaded kernel_mul_row                        0x13881d6a0
ggml_metal_init: loaded kernel_scale                          0x13881e100
ggml_metal_init: loaded kernel_silu                           0x13881e940
ggml_metal_init: loaded kernel_relu                           0x13881c480
ggml_metal_init: loaded kernel_gelu                           0x13881f240
ggml_metal_init: loaded kernel_soft_max                       0x138820280
ggml_metal_init: loaded kernel_diag_mask_inf                  0x1388214e0
ggml_metal_init: loaded kernel_get_rows_f16                   0x138821740
ggml_metal_init: loaded kernel_get_rows_q4_0                  0x1073044f0
ggml_metal_init: loaded kernel_get_rows_q4_1                  0x1073050a0
ggml_metal_init: loaded kernel_get_rows_q2_K                  0x107305610
ggml_metal_init: loaded kernel_get_rows_q3_K                  0x107305f30
ggml_metal_init: loaded kernel_get_rows_q4_K                  0x107306980
ggml_metal_init: loaded kernel_get_rows_q5_K                  0x1073072d0
ggml_metal_init: loaded kernel_get_rows_q6_K                  0x107307ad0
ggml_metal_init: loaded kernel_rms_norm                       0x107308540
ggml_metal_init: loaded kernel_norm                           0x107309640
ggml_metal_init: loaded kernel_mul_mat_f16_f32                0x10730a240
ggml_metal_init: loaded kernel_mul_mat_q4_0_f32               0x10730adc0
ggml_metal_init: loaded kernel_mul_mat_q4_1_f32               0x10730b790
ggml_metal_init: loaded kernel_mul_mat_q2_K_f32               0x10730c140
ggml_metal_init: loaded kernel_mul_mat_q3_K_f32               0x10730cc50
ggml_metal_init: loaded kernel_mul_mat_q4_K_f32               0x10730d580
ggml_metal_init: loaded kernel_mul_mat_q5_K_f32               0x10730df60
ggml_metal_init: loaded kernel_mul_mat_q6_K_f32               0x10730e930
ggml_metal_init: loaded kernel_rope                           0x10730eb90
ggml_metal_init: loaded kernel_alibi_f32                      0x10730f5c0
ggml_metal_init: loaded kernel_cpy_f32_f16                    0x107310140
ggml_metal_init: loaded kernel_cpy_f32_f32                    0x138822560
ggml_metal_init: loaded kernel_cpy_f16_f16                    0x130372a50
ggml_metal_init: recommendedMaxWorkingSetSize = 10922.67 MB
ggml_metal_init: hasUnifiedMemory             = true
ggml_metal_init: maxTransferRate              = built-in GPU
whisper_init_from_file: max tensor size =    26.12 MB
ggml_metal_add_buffer: allocated 'data            ' buffer, size =   176.00 MB, (  176.45 / 10922.67)
ggml_metal_add_buffer: allocated 'eval            ' buffer, size =    56.00 MB, (  232.45 / 10922.67)
ggml_metal_add_buffer: allocated 'kvself          ' buffer, size =    16.00 MB, (  248.45 / 10922.67)
ggml_metal_add_buffer: allocated 'kvcross         ' buffer, size =    53.00 MB, (  301.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr0            ' buffer, size =   120.00 MB, (  421.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr1            ' buffer, size =    36.00 MB, (  457.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr2            ' buffer, size =     6.00 MB, (  463.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr3            ' buffer, size =     6.00 MB, (  469.45 / 10922.67)

system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | COREML = 0 | OPENVINO = 0 | METAL = 1 | K_QUANTS = 1 | 

main: processing './samples/jfk.wav' (176000 samples, 11.0 sec), 4 threads, 1 processors, lang = en, task = transcribe, timestamps = 1 ...


[00:00:00.000 --> 00:00:07.000]   And! So, my fellow Americans, ask not what your country can do!
[00:00:07.000 --> 00:00:11.000]   Ask what! you can do for your country!


whisper_print_timings:     load time =   150.77 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    34.22 ms
whisper_print_timings:   sample time =    11.23 ms /    30 runs (    0.37 ms per run)
whisper_print_timings:   encode time =  1105.81 ms /     1 runs ( 1105.81 ms per run)
whisper_print_timings:   decode time =   210.00 ms /    30 runs (    7.00 ms per run)
whisper_print_timings:    total time =  1559.63 ms

``` 
</details>

<details><summary>Medium Q5_K CPU</summary>

```
whisper_init_from_file_no_state: loading model from 'models/ggml-medium_q5_k.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1024
whisper_model_load: n_audio_head  = 16
whisper_model_load: n_audio_layer = 24
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1024
whisper_model_load: n_text_head   = 16
whisper_model_load: n_text_layer  = 24
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 13
whisper_model_load: qntvr         = 2
whisper_model_load: type          = 4
whisper_model_load: mem required  =  975.00 MB (+   43.00 MB per decoder)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: model ctx     =  513.23 MB
whisper_model_load: model size    =  512.77 MB
whisper_init_state: kv self size  =   42.00 MB
whisper_init_state: kv cross size =  140.62 MB

system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | COREML = 0 | OPENVINO = 0 | METAL = 0 | K_QUANTS = 1 | 

main: processing './samples/jfk.wav' (176000 samples, 11.0 sec), 4 threads, 1 processors, lang = en, task = transcribe, timestamps = 1 ...


[00:00:00.000 --> 00:00:07.800]   And so, my fellow Americans, ask not what your country can do for you.
[00:00:07.800 --> 00:00:10.800]   Ask what you can do for your country.


whisper_print_timings:     load time =   182.82 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    34.30 ms
whisper_print_timings:   sample time =    11.66 ms /    30 runs (    0.39 ms per run)
whisper_print_timings:   encode time =  3035.53 ms /     1 runs ( 3035.53 ms per run)
whisper_print_timings:   decode time =   344.22 ms /    30 runs (   11.47 ms per run)
whisper_print_timings:    total time =  3630.39 ms
``` 
</details>

<details><summary>Medium Q5_K GPU</summary>

```
whisper_init_from_file_no_state: loading model from 'models/ggml-medium_q5_k.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1024
whisper_model_load: n_audio_head  = 16
whisper_model_load: n_audio_layer = 24
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1024
whisper_model_load: n_text_head   = 16
whisper_model_load: n_text_layer  = 24
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 13
whisper_model_load: qntvr         = 2
whisper_model_load: type          = 4
whisper_model_load: mem required  =  975.00 MB (+   43.00 MB per decoder)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: model ctx     =  513.23 MB
whisper_model_load: model size    =  512.77 MB
whisper_init_state: kv self size  =   42.00 MB
whisper_init_state: kv cross size =  140.62 MB
ggml_metal_init: allocating
ggml_metal_init: using MPS
ggml_metal_init: loaded kernel_add                            0x12462d840
ggml_metal_init: loaded kernel_add_row                        0x12462fa00
ggml_metal_init: loaded kernel_mul                            0x12462ee60
ggml_metal_init: loaded kernel_mul_row                        0x124630480
ggml_metal_init: loaded kernel_scale                          0x124630ee0
ggml_metal_init: loaded kernel_silu                           0x1246316f0
ggml_metal_init: loaded kernel_relu                           0x12462f260
ggml_metal_init: loaded kernel_gelu                           0x124632030
ggml_metal_init: loaded kernel_soft_max                       0x124632f60
ggml_metal_init: loaded kernel_diag_mask_inf                  0x1246342d0
ggml_metal_init: loaded kernel_get_rows_f16                   0x124634530
ggml_metal_init: loaded kernel_get_rows_q4_0                  0x124635740
ggml_metal_init: loaded kernel_get_rows_q4_1                  0x124634bc0
ggml_metal_init: loaded kernel_get_rows_q2_K                  0x124634e20
ggml_metal_init: loaded kernel_get_rows_q3_K                  0x124636140
ggml_metal_init: loaded kernel_get_rows_q4_K                  0x124636ac0
ggml_metal_init: loaded kernel_get_rows_q5_K                  0x124637280
ggml_metal_init: loaded kernel_get_rows_q6_K                  0x124637b90
ggml_metal_init: loaded kernel_rms_norm                       0x124638670
ggml_metal_init: loaded kernel_norm                           0x124639790
ggml_metal_init: loaded kernel_mul_mat_f16_f32                0x124639b90
ggml_metal_init: loaded kernel_mul_mat_q4_0_f32               0x12463ad90
ggml_metal_init: loaded kernel_mul_mat_q4_1_f32               0x12463b750
ggml_metal_init: loaded kernel_mul_mat_q2_K_f32               0x12463c140
ggml_metal_init: loaded kernel_mul_mat_q3_K_f32               0x12463cc60
ggml_metal_init: loaded kernel_mul_mat_q4_K_f32               0x12463d720
ggml_metal_init: loaded kernel_mul_mat_q5_K_f32               0x12463df70
ggml_metal_init: loaded kernel_mul_mat_q6_K_f32               0x12463e940
ggml_metal_init: loaded kernel_rope                           0x12463f3e0
ggml_metal_init: loaded kernel_alibi_f32                      0x12463ffa0
ggml_metal_init: loaded kernel_cpy_f32_f16                    0x124640e30
ggml_metal_init: loaded kernel_cpy_f32_f32                    0x1246419e0
ggml_metal_init: loaded kernel_cpy_f16_f16                    0x124638fb0
ggml_metal_init: recommendedMaxWorkingSetSize = 10922.67 MB
ggml_metal_init: hasUnifiedMemory             = true
ggml_metal_init: maxTransferRate              = built-in GPU
whisper_init_from_file: max tensor size =    34.82 MB
ggml_metal_add_buffer: allocated 'data            ' buffer, size =   540.00 MB, (  540.45 / 10922.67)
ggml_metal_add_buffer: allocated 'eval            ' buffer, size =    74.00 MB, (  614.45 / 10922.67)
ggml_metal_add_buffer: allocated 'kvself          ' buffer, size =    43.00 MB, (  657.45 / 10922.67)
ggml_metal_add_buffer: allocated 'kvcross         ' buffer, size =   141.00 MB, (  798.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr0            ' buffer, size =   158.00 MB, (  956.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr1            ' buffer, size =    48.00 MB, ( 1004.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr2            ' buffer, size =     7.00 MB, ( 1011.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr3            ' buffer, size =     7.00 MB, ( 1018.45 / 10922.67)

system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | COREML = 0 | OPENVINO = 0 | METAL = 1 | K_QUANTS = 1 | 

main: processing './samples/jfk.wav' (176000 samples, 11.0 sec), 4 threads, 1 processors, lang = en, task = transcribe, timestamps = 1 ...


[00:00:00.000 --> 00:00:07.800]   And so, my fellow Americans, ask not what your country can do for you.
[00:00:07.800 --> 00:00:10.800]   Ask what you can do for your country.


whisper_print_timings:     load time =   179.97 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    33.94 ms
whisper_print_timings:   sample time =    13.76 ms /    30 runs (    0.46 ms per run)
whisper_print_timings:   encode time =  2981.41 ms /     1 runs ( 2981.41 ms per run)
whisper_print_timings:   decode time =   432.81 ms /    30 runs (   14.43 ms per run)
whisper_print_timings:    total time =  3695.76 ms
ggml_metal_free: deallocating
``` 
</details>

<details><summary>Large Q5_K CPU</summary>

```
whisper_init_from_file_no_state: loading model from 'models/ggml-large_q5_k.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1280
whisper_model_load: n_audio_head  = 20
whisper_model_load: n_audio_layer = 32
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1280
whisper_model_load: n_text_head   = 20
whisper_model_load: n_text_layer  = 32
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 13
whisper_model_load: qntvr         = 2
whisper_model_load: type          = 5
whisper_model_load: mem required  = 1686.00 MB (+   71.00 MB per decoder)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: model ctx     = 1029.58 MB
whisper_model_load: model size    = 1028.97 MB
whisper_init_state: kv self size  =   70.00 MB
whisper_init_state: kv cross size =  234.38 MB

system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | COREML = 0 | OPENVINO = 0 | METAL = 0 | K_QUANTS = 1 | 

main: processing './samples/jfk.wav' (176000 samples, 11.0 sec), 4 threads, 1 processors, lang = en, task = transcribe, timestamps = 1 ...


[00:00:00.000 --> 00:00:11.000]   And so my fellow Americans, ask not what your country can do for you, ask what you can do for your country.


whisper_print_timings:     load time =   479.21 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    40.84 ms
whisper_print_timings:   sample time =    10.74 ms /    27 runs (    0.40 ms per run)
whisper_print_timings:   encode time =  5745.24 ms /     1 runs ( 5745.24 ms per run)
whisper_print_timings:   decode time =   696.49 ms /    27 runs (   25.80 ms per run)
whisper_print_timings:    total time =  7029.23 ms
``` 
</details>

<details><summary>Large Q5_K GPU</summary>

```
whisper_init_from_file_no_state: loading model from 'models/ggml-large_q5_k.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51865
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1280
whisper_model_load: n_audio_head  = 20
whisper_model_load: n_audio_layer = 32
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1280
whisper_model_load: n_text_head   = 20
whisper_model_load: n_text_layer  = 32
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 13
whisper_model_load: qntvr         = 2
whisper_model_load: type          = 5
whisper_model_load: mem required  = 1686.00 MB (+   71.00 MB per decoder)
whisper_model_load: adding 1608 extra tokens
whisper_model_load: model ctx     = 1029.58 MB
whisper_model_load: model size    = 1028.97 MB
whisper_init_state: kv self size  =   70.00 MB
whisper_init_state: kv cross size =  234.38 MB
ggml_metal_init: allocating
ggml_metal_init: using MPS
ggml_metal_init: loaded kernel_add                            0x149ca6390
ggml_metal_init: loaded kernel_add_row                        0x149ca8d40
ggml_metal_init: loaded kernel_mul                            0x149ca8120
ggml_metal_init: loaded kernel_mul_row                        0x149ca97c0
ggml_metal_init: loaded kernel_scale                          0x149caa230
ggml_metal_init: loaded kernel_silu                           0x149caaa70
ggml_metal_init: loaded kernel_relu                           0x149ca8520
ggml_metal_init: loaded kernel_gelu                           0x149cab3a0
ggml_metal_init: loaded kernel_soft_max                       0x149cac2e0
ggml_metal_init: loaded kernel_diag_mask_inf                  0x149cad650
ggml_metal_init: loaded kernel_get_rows_f16                   0x149cad8b0
ggml_metal_init: loaded kernel_get_rows_q4_0                  0x149caeb10
ggml_metal_init: loaded kernel_get_rows_q4_1                  0x155f597c0
ggml_metal_init: loaded kernel_get_rows_q2_K                  0x155f5a550
ggml_metal_init: loaded kernel_get_rows_q3_K                  0x155f5aa70
ggml_metal_init: loaded kernel_get_rows_q4_K                  0x155f5b4a0
ggml_metal_init: loaded kernel_get_rows_q5_K                  0x1498774f0
ggml_metal_init: loaded kernel_get_rows_q6_K                  0x149878260
ggml_metal_init: loaded kernel_rms_norm                       0x149878870
ggml_metal_init: loaded kernel_norm                           0x14987bab0
ggml_metal_init: loaded kernel_mul_mat_f16_f32                0x14987c6b0
ggml_metal_init: loaded kernel_mul_mat_q4_0_f32               0x149e04a30
ggml_metal_init: loaded kernel_mul_mat_q4_1_f32               0x149e05020
ggml_metal_init: loaded kernel_mul_mat_q2_K_f32               0x155f5be20
ggml_metal_init: loaded kernel_mul_mat_q3_K_f32               0x155f5cf80
ggml_metal_init: loaded kernel_mul_mat_q4_K_f32               0x14987c910
ggml_metal_init: loaded kernel_mul_mat_q5_K_f32               0x14987d620
ggml_metal_init: loaded kernel_mul_mat_q6_K_f32               0x14987dba0
ggml_metal_init: loaded kernel_rope                           0x155e192c0
ggml_metal_init: loaded kernel_alibi_f32                      0x155e19f00
ggml_metal_init: loaded kernel_cpy_f32_f16                    0x155e1af30
ggml_metal_init: loaded kernel_cpy_f32_f32                    0x14987f5c0
ggml_metal_init: loaded kernel_cpy_f16_f16                    0x149878d10
ggml_metal_init: recommendedMaxWorkingSetSize = 10922.67 MB
ggml_metal_init: hasUnifiedMemory             = true
ggml_metal_init: maxTransferRate              = built-in GPU
whisper_init_from_file: max tensor size =    43.53 MB
ggml_metal_add_buffer: allocated 'data            ' buffer, size =  1081.00 MB, ( 1081.45 / 10922.67)
ggml_metal_add_buffer: allocated 'eval            ' buffer, size =    94.00 MB, ( 1175.45 / 10922.67)
ggml_metal_add_buffer: allocated 'kvself          ' buffer, size =    71.00 MB, ( 1246.45 / 10922.67)
ggml_metal_add_buffer: allocated 'kvcross         ' buffer, size =   235.00 MB, ( 1481.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr0            ' buffer, size =   198.00 MB, ( 1679.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr1            ' buffer, size =    60.00 MB, ( 1739.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr2            ' buffer, size =     9.00 MB, ( 1748.45 / 10922.67)
ggml_metal_add_buffer: allocated 'scr3            ' buffer, size =     9.00 MB, ( 1757.45 / 10922.67)

system_info: n_threads = 4 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | COREML = 0 | OPENVINO = 0 | METAL = 1 | K_QUANTS = 1 | 

main: processing './samples/jfk.wav' (176000 samples, 11.0 sec), 4 threads, 1 processors, lang = en, task = transcribe, timestamps = 1 ...


[00:00:00.000 --> 00:00:11.000]   And so my fellow Americans, ask not what your country can do for you, ask what you can do for your country.


whisper_print_timings:     load time =   657.45 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    34.17 ms
whisper_print_timings:   sample time =    11.60 ms /    27 runs (    0.43 ms per run)
whisper_print_timings:   encode time =  6002.63 ms /     1 runs ( 6002.63 ms per run)
whisper_print_timings:   decode time =   766.03 ms /    27 runs (   28.37 ms per run)
whisper_print_timings:    total time =  7651.89 ms
ggml_metal_free: deallocating
``` 
</details>
